Faster MmapAllocator::allocate()

oerling opened the pull request:
Adds a second level bitmap for finding free pages backed by memory in
size classes. Checks allocated bitmaps 256 bits at a time with
SIMD. The top level bitmap has one bit for every 512 bits in
allocation and mapping bitmaps.

Adds consistency checks for the new mappedFreeLookup_ bitmap.

tanjialiang commented on velox/common/memory/MmapAllocator.cpp:
```diff
@@ -384,31 +425,112 @@ bool MmapAllocator::SizeClass::allocateLocked(
+      allocateAny(cursor, numPagesToGo, *numUnmapped, out);
+      numAllocatedUnmapped_ += previousToGo - numPagesToGo;
+      if (numPagesToGo == 0) {
+        clockHand_ = previousCursor;
```
It seems like we did not update clockHand_ in previous version. I wonder this contributes to performance gain or not.

tanjialiang commented on velox/common/memory/MmapAllocator.cpp:
```diff
@@ -384,31 +425,112 @@ bool MmapAllocator::SizeClass::allocateLocked(
+      int previousToGo = numPagesToGo;
+      assert(numUnmapped != nullptr);
+      allocateAny(cursor, numPagesToGo, *numUnmapped, out);
+      numAllocatedUnmapped_ += previousToGo - numPagesToGo;
```
How are we sure all newly given class pages are unmapped?

tanjialiang commented on velox/common/memory/MmapAllocator.cpp:
```diff
@@ -384,31 +425,112 @@ bool MmapAllocator::SizeClass::allocateLocked(
+  }
+}
+
+namespace {
```
-------------Stopped here. Continuing next time-------------

oerling commented:
Updated the files. Here’s replies. We really want to use implicit cast to bool over == 0 for existence with counters or pointers.

 

 

 

 

Is there any specific reasons we want to call it words instead of bytes? Is it instruction related? Can we get away of calling them bytes instead?

 

 

 

 

A:  This is a little time critical so we want to check 256 bits at a time. Because the closest native C type is uint64_t we talk about words. Word is better than byte because we never process these arrays as bytes.  The semantics are all the same. The bit order is the machine little endian order. xsimd::batch<uint64_t> is consistently used. This is 4 adjacent elements of the arrays.

 

 

 

I wonder if we should extract the 64 out as a constant in this class, as many other places are using it as well. For consistency purposes. Maybe

constexpr int32_t kWordSize = sizeof(uint64_t);

or something

 

A: / 64  is a lot more readable than / sizeof(int64_t) * 8 or / kBitsInInt64.

 

We always have bitmaps made of uint64_t so there is consistent practice. 64 is the minimal idiom so one does not have to think about what it means.

 

Isn't this going to overshoot when i = pageAllocated_.size() - 1

 

 

 

A: we assert that the vectors have kSindTail at construction.

 

 

Can we do if (mappedFreeInGroup == 0) ?

 

 

A: The idiom !x is conventionally used to mean none when talking about a count.  So this is consistent with precedent. If we checked for multiple we'd say (x % 64) == 0 as opposed to !(x % 64).

 

 

 

Should we throw to surface issues earlier?

 

 

A: This function should not throw to start with. It is supposed to list the discrepancies. Note that this is not a run time integrity check because to check consistency between classes you'd need a global mutex and there is not one. But in testing this is supposed to finish. So time of detection is not a factor.

 

Generally inconsistencies here are a process killing situation because these signal some intractable corruption even though the gizmo can keep running for a long time.

 

 

Also here shouldn't we throw?

 

 

 

A: No. We should list what's broken and let the caller fail.

 

Trying to understand, why are we comparing number of units (numAllocated) with number of pages (considerMappedOnly) ?

 

 

A: considerMappedOnly a count of class pages (units). Updated the declaration.

 

   size_t numWords = pageAllocated_.size();

Could really be numClassPages

 

 

A: No, this is the number of words in the bitmaps. The class page count is 64 * that.

 

+      assert(numUnmapped != nullptr);

Assert here is not needed as we have already checked before that numUnmapped is not nullptr

 

A: Correct, but fbcode lint whats this just the same. Let's see.

 

It seems like we did not update clockHand_ in previous version. I wonder this contributes to performance gain or not.

 

 

A: Since this applies to only cases of allocating where there is no backing page, this sis a slow path anyway. Updating clockhand will just make it so the pages drift along the address range of the size class.  This equalizes times as opposed to for example allocating the lowest available. There is a minor fault anyway on first write so time is not very important.

 

 

+      numAllocatedUnmapped_ += previousToGo - numPagesToGo;

How are we sure all newly given class pages are unmapped?

 

 

 

A: We only come here after all the mapped free pages of the class are used up.

 

Ah okay they are both class pages. I was confused. I wonder if it makes sense to rename the fields inside size class that represent class pages to somethingClassPages instead of somethingPages

—

 

A: I think the type name can serve as comment for this.

tanjialiang commented on velox/common/memory/MmapAllocator.cpp:
```diff
@@ -404,31 +447,111 @@ bool MmapAllocator::SizeClass::allocateLocked(
+  }
+  ClassPageCount ignore = 0;
+  checkConsistency(ignore, ignore);
+  LOG(ERROR) << "MMAPL: Inconsistent mapped free lookup class " << unitSize_;
```
To your comments somewhere else but related here:

> Generally inconsistencies here are a process killing situation because these signal some intractable corruption even though the gizmo can keep running for a long time.

So here the checkConsistency is called in runtime while doing allocation, instead of during testing. You also mentioned time of detection is not a factor. Wouldn't it be better to fail fast for easier debugging? This kind of issues cannot be detected by ASAN because we are using our own mapped memory. Would like to know your thoughts.

tanjialiang commented on velox/common/memory/MmapAllocator.cpp:
```diff
@@ -544,6 +667,7 @@ MachinePageCount MmapAllocator::SizeClass::free(
       }
       if (bits::isBitSet(pageMapped_.data(), page)) {
         ++numMappedFreePages_;
+        markMappedFree(page);
```
Why are we freeing the lookup bit when only one class page is freed in a range of class pages represented by a lookup bit?

tanjialiang commented:
Thanks for working on this. Could we import it to fbcode for more signals?

mbasmanova reviewed (commented):
@oerling How much faster is the new implementation? Do you have a benchmark?

oerling commented:
This is ~20x faster. In running Velox_frragmentation_benchmark, allocation was around 10% of profile, now it is ~0.5%. Stands to reason since we do one 256 bit check as opposed to up to 2048 64 bit word checks.

 

This is somewhat meaningful if allocating vector Buffers from MmapAllocator. This should be done for convergence of accounting and keeping malloc footprint low. It is not very actionable if you see high allocation and this is mostly Buffers from malloc because you can’t compensate for these by evicting cache, for example. Also, the only shrinkage by memory arbitration is in mappedMemory, so vectors should come from there, otherwise you can’t use this for managing vector allocation peaks (outlier row sizes, e.g. exceptionally large maps and map intermediate results).  

 

We probably get there as soon as running larger queries.
